### PR TITLE
Allow shards to be allocated if leftover shard from different index exists.

### LINKED ISSUE #10677: Ensure we can recover cleanly into a leftover shard directory from a different index
this is a followup issue for #10461 where we now throw an exception if we try to allocate a shard where there already exists such a shards with a non-matching index UUID. This is an edge case but we should be able to recover cleanly in such a case an remove that leftover shard directory.

### PR DESCRIPTION
**Pull request by s1monw**
If an index name is reused but a leftover shard still exists on any node
we fail repeatedly to allocate the shard since we now check the index UUID
before reusing data. This commit allows to recover even if there is such a
leftover shard by trying to archive the shard into a seperate directory.

Closes #10677

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/shard/ShardPath.java [hunk 1]
```diff
@@ -118,6 +118,32 @@ public static ShardPath loadShardPath(ESLogger logger, NodeEnvironment env, Shar
...
+                    logger.warn("{} found shard on path: [{}] with a different index UUID - this shard seems to be leftover from a different index with the same name.", shardId, path);
+                    assert Files.isDirectory(path) : path + " is not a directory";
+                    final Path target = path.resolveSibling("archived_" + path.getFileName().toString() + "_" + load.indexUUID);
+                    try {
+                        Files.move(path, target); // rename directory
+                    } catch (IOException ex) {
+                        logger.warn("{} failed to archive shard from an index with a different UUID into {}", ex, shardId, target);
+                        throw ex;
+                    }
+                    logger.warn("{} archived shard from an index with a different UUID into {}, Remove the leftover shard if it's not needed anymore", shardId, target);
```
Comment (clintongormley) [inline core/src/main/java/org/elasticsearch/index/shard/ShardPath.java]: Possibly "Delete the leftover shard manually..."?

**Comment by s1monw** [comment]
@bleskes can you take a look if you have some time

**Comment by bleskes** [comment]
If I understand correctly, our primary concern here is to not delete old indices by mistake because a node joins a cluster it wasn't supposed to or some other weird case. In this case we shouldn't archive into the same index folder as we do now, but rather into another index level folder. If we keep it locally we might run the risk of cleaning up the archived shard folder by mistake when all shards of the active index are moved to another node.

The flip of this is that if we put it into another index folder, we should change the meta data as well (to make it consistent) which also means it will be picked up by the dangling indices, surfacing it up to the user and also allowing them to deal with it using the API (delete it).

**Comment by s1monw** [comment]
you are right, I think the simplest solution is to just delete the stuff there and be done with it. its such a conrner case I don't want to overdesign for cornercases and bogus like this.

**Comment by bleskes** [comment]
++ to maintain old behavior and clean it up for now.

**Pull request closed without merging**